### ISSUE TITLE
RegExpMatchArray has incorrect type definition

### ISSUE THREAD

**Issue by kirilloid** [Microsoft/TypeScript#17053]
**TypeScript Version:**  2.4.1
The `strict` option is enabled.

**Code**

```ts
const match = "a".match(/(a)?(b)?/);
if (match !== null) {
    const [, a, b] = match;
    alert(b.toUpperCase());
}
```

**Expected behavior:**
The compiler should warn that `b` might be `undefined`.

**Actual behavior:**
No compile errors, and run-time error.

`interface RegExpMatchArray extends Array<string>` should be `Array<string|undefined>`

**Comment by RyanCavanaugh** [comment]
This is intentionally less strict to avoid unnecessary friction with non-optional groups (which are by far more common IME). Maybe we'd have gone a different direction if we started with strict null checks from the beginning, but this isn't the sort of breaking change we would want to do now.

**Comment by AnyhowStep** [comment]
Kinda' old and everyone's probably already moved on but I thought I'd just put this here,

### Less Intrusive

You could have a `global.d.ts` of sorts, and throw this in,

```
interface RegExpExecArrayOptionalItems extends Array<string|undefined> {
    index: number;
    input: string;
}
```

Then, where ever else,

```
const match : RegExpExecArrayOptionalItems|null = /^(\d+)?.+$/.exec("hahaha");
//match will be RegExpExecArrayOptionalItems
//match[1] will be undefined
//The type checker will work as intended
```

But, as you can see, you have to manually declare the type of `match`.

-----

### More Intrusive

I don't really like this one

You could have a `global.d.ts` of sorts, and throw this in,

```
interface RegExpExecArrayOptionalItems extends Array<string|undefined> {
    index: number;
    input: string;
}
interface RegExp {
    /**
      * Executes a search on a string using a regular expression pattern, and returns an array containing the results of that search.
      * @param string The String object or string literal on which to perform the search.
      */
    exec(string: string): RegExpExecArrayOptionalItems | null;
}
```

Then, where ever else,

```
const match = /^(\d+)$/.exec("123");
//match will be RegExpExecArrayOptionalItems
//match[1] will be string|undefined, even though our regex says to match only if the group exists
```

But, as you can see, even groups that are "guaranteed" to exist will be marked as potentially undefined
